snap,snappy: refactor to remove "Install" from the Part interface

mvo5 opened the pull request:
This branch removes "Install" from the part interface. Install of snaps is part of the overlord. The downloading is part of the store. There is a new helper to glue it together.

zyga commented on snappy/install.go:
```diff
@@ -45,13 +45,34 @@ const (
+func installRemote(mStore *SnapUbuntuStoreRepository, remoteSnap *RemoteSnapPart, flags InstallFlags, meter progress.Meter) (string, error) {
+	downloadedSnap, err := mStore.Download(remoteSnap, meter)
+	if err != nil {
+		return "", fmt.Errorf("can not download %s: %s", remoteSnap.Name(), err)
```
s/can not/cannot/ or can't. To me personally "can not" looks a bit odd.

chipaca replied:
this is probably why we've been using “unable to X” instead of “can't X” or other forms.

niemeyer replied:
We've actually been using "cannot", at least in the reviews I've been in. Seems like the better choice for being shorter and easier to read/write/say. If we change this, let's please sync on it so we're aligned.

zyga commented:
Looks good. I've added one comment on error message wording but feel free to ignore that.

chipaca commented on snappy/install.go:
```diff
@@ -100,15 +121,19 @@ func Update(name string, flags InstallFlags, meter progress.Meter) ([]Part, erro
-	updates, err := ListUpdates()
+	updates, err := mStore.Updates()
+	if err != nil {
+		return nil, fmt.Errorf("can not get updates: %s", err)
```
same here wrt `can not`

chipaca commented:
:+1:

nits around error phrasing.

niemeyer commented on snappy/snap_remote_repo.go:
```diff
@@ -383,3 +386,61 @@ func (s *SnapUbuntuStoreRepository) Updates() (parts []Part, err error) {
 }
+
+// Download downloads the given snap and returns the filename of the local
+// version
```
```
// Download downloads the given snap and returns its filename.
// The file is saved in temporary storage, and should be removed
// after use to prevent the disk from running out of space.
```

niemeyer commented on snappy/snap_remote_repo.go:
```diff
@@ -383,3 +386,61 @@ func (s *SnapUbuntuStoreRepository) Updates() (parts []Part, err error) {
+	if err != nil {
+		return "", err
+	}
+	defer func() {
```
if cerr := w.Close(); cerr != nil && err == nil {
        err = cerr
}

niemeyer commented on snappy/snap_remote_repo.go:
```diff
@@ -383,3 +386,61 @@ func (s *SnapUbuntuStoreRepository) Updates() (parts []Part, err error) {
+			os.Remove(w.Name())
+		}
+	}()
+	defer w.Close()
```
... and this can then go away.

niemeyer commented on snappy/snap_remote_repo.go:
```diff
@@ -383,3 +386,61 @@ func (s *SnapUbuntuStoreRepository) Updates() (parts []Part, err error) {
+	setUbuntuStoreHeaders(req)
+
+	if err := download(remoteSnap.Name(), w, req, pbar); err != nil {
+		return "", err
```
Oops.. err here does not point to the err in the defer, because of the ":=", so the file wouldn't get removed. Subtle.

chipaca replied:
umm... no, i don't think that's how it works. See https://play.golang.org/p/RqwfZCAiaq

mvo5 replied:
I wrote a test to ensure that I got this right :) thanks to both of you!

niemeyer commented on snappy/snap_remote_repo.go:
```diff
@@ -383,3 +386,61 @@ func (s *SnapUbuntuStoreRepository) Updates() (parts []Part, err error) {
+		return "", err
+	}
+
+	return w.Name(), w.Sync()
```
I'm not 100% sure that syncing covers has all the important low-level semantics of closing the file. For the typical case, probably, but it sounds so easy to have it clearly right.

There's also another issue about the w.Name() return here. The defer removes the file, but the call site still gets the name pointing to a non-existent file. Instead, this function should be declared with named results (e.g. `(path string, err error)`), and the path reset after the file is removed on errors.

chipaca replied:
syncing does not do the whole dance. Unfortunately the atomic file operation we have takes a []byte, instead of implementing a writer. Maybe we should refactor it a bit and then use that here; having a WriterComitter interface kind.

chipaca replied:
in any case note that this is not a change from this branch, and shouldn't affect it imho.

mvo5 commented:
Thanks for the excellent review and thanks for spotting the issues. I addressed them now and added tests to ensure that we won't hit regressions here.

chipaca commented:
retest this please
